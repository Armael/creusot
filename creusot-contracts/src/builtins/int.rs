use std::ops::*;

#[rustc_diagnostic_item = "creusot_int"]
pub struct Int;

impl PartialEq for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "eq_int"]
    fn eq(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "ne_int"]
    fn ne(&self, _: &Int) -> bool {
        panic!()
    }
}

impl PartialOrd for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "le_int"]
    fn le(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "ge_int"]
    fn ge(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "lt_int"]
    fn lt(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "gt_int"]
    fn gt(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    fn partial_cmp(&self, _: &Self) -> Option<std::cmp::Ordering> {
        panic!()
    }
}

impl From<i32> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "i32_to_int"]
    fn from(_: i32) -> Self {
        panic!()
    }
}

impl From<u32> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "u32_to_int"]
    fn from(_: u32) -> Self {
        panic!()
    }
}

impl From<usize> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "usize_to_int"]
    fn from(_: usize) -> Self {
        panic!()
    }
}

impl Add<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "add_int"]
    fn add(self, _: Int) -> Self {
        panic!()
    }
}

impl Sub<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "sub_int"]
    fn sub(self, _: Int) -> Self {
        panic!()
    }
}

impl Mul<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "mul_int"]
    fn mul(self, _: Int) -> Self {
        panic!()
    }
}

impl Div<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "div_int"]
    fn div(self, _: Int) -> Self {
        panic!()
    }
}

impl Rem<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "rem_int"]
    fn rem(self, _: Int) -> Self {
        panic!()
    }
}

impl Neg for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "neg_int"]
    fn neg(self) -> Self {
        panic!()
    }
}
