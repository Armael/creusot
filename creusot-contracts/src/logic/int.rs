use super::model::*;
use crate as creusot_contracts;
use crate::logic::*;
use creusot_contracts_proc::*;

use std::ops::*;

#[rustc_diagnostic_item = "creusot_int"]
pub struct Int(*mut ());

impl WellFounded for Int {}

impl Int {
    #[trusted]
    #[logic]
    #[creusot::builtins = "int.Power.power"]
    pub fn pow(self, _: Int) -> Int {
        absurd
    }
}

impl From<u8> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "prelude.UInt8.to_int"]
    fn from(_: u8) -> Self {
        absurd
    }
}
impl Model for u8 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "prelude.UInt8.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<u32> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.UInt32.to_int"]
    fn from(_: u32) -> Self {
        absurd
    }
}
impl Model for u32 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.UInt32.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<u64> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn from(_: u64) -> Self {
        absurd
    }
}
impl Model for u64 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<i64> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.Int64.to_int"]
    fn from(_: i64) -> Self {
        absurd
    }
}
impl Model for i64 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.Int64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<i32> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.Int32.to_int"]
    fn from(_: i32) -> Self {
        absurd
    }
}
impl Model for i32 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.Int32.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<usize> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn from(_: usize) -> Self {
        absurd
    }
}
impl Model for usize {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<isize> for Int {
    #[logic]
    #[trusted]
    #[rustc_diagnostic_item = "isize_to_int"]
    fn from(_: isize) -> Self {
        absurd
    }
}
impl Model for isize {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.Int64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl Add<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "add_int"]
    fn add(self, _: Int) -> Self {
        panic!()
    }
}

impl Sub<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "sub_int"]
    fn sub(self, _: Int) -> Self {
        panic!()
    }
}

impl Mul<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "mul_int"]
    fn mul(self, _: Int) -> Self {
        panic!()
    }
}

impl Div<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "div_int"]
    fn div(self, _: Int) -> Self {
        panic!()
    }
}

impl Rem<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "rem_int"]
    fn rem(self, _: Int) -> Self {
        panic!()
    }
}

impl Neg for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "neg_int"]
    fn neg(self) -> Self {
        panic!()
    }
}
