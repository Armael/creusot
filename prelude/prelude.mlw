module CreusotPrelude
  type borrowed 'a = { current : 'a; final: 'a }

  let function ( *_ ) x = x.current
  let function ( ^_ ) x = x.final

  val borrow_mut (a : 'a) : borrowed 'a
    ensures { * result = a }
end
