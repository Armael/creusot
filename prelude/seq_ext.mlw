module SeqExt
use seq.Seq

function subsequence (s : seq 't) (i : int) (j : int) : seq 't = s[i..j]

function mk_empty () : seq 't = Seq.empty
end