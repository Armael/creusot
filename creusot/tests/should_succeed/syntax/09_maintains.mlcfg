
module C09Maintains_A_Type
  type c09maintains_a_type  =
    | C09Maintains_A_Type
    
end
module C09Maintains_Impl0_Invariant_Interface
  use mach.int.Int
  use mach.int.UInt64
  use C09Maintains_A_Type
  predicate invariant' (self : C09Maintains_A_Type.c09maintains_a_type) (b : bool) (c : uint64)
end
module C09Maintains_Impl0_Invariant
  use mach.int.Int
  use mach.int.UInt64
  use C09Maintains_A_Type
  predicate invariant' [#"../09_maintains.rs" 13 4 13 47] (self : C09Maintains_A_Type.c09maintains_a_type) (b : bool) (c : uint64)
    
   =
    [#"../09_maintains.rs" 14 8 14 12] true
end
module C09Maintains_Impl0_Inv2_Interface
  use mach.int.Int
  use C09Maintains_A_Type
  predicate inv2 (self : C09Maintains_A_Type.c09maintains_a_type) (b : int)
end
module C09Maintains_Impl0_Inv2
  use mach.int.Int
  use C09Maintains_A_Type
  predicate inv2 [#"../09_maintains.rs" 18 4 18 33] (self : C09Maintains_A_Type.c09maintains_a_type) (b : int) =
    [#"../09_maintains.rs" 19 8 19 12] true
end
module C09Maintains_OtherInv_Interface
  use C09Maintains_A_Type
  predicate other_inv (a : C09Maintains_A_Type.c09maintains_a_type) (b : bool)
end
module C09Maintains_OtherInv
  use C09Maintains_A_Type
  predicate other_inv [#"../09_maintains.rs" 24 0 24 35] (a : C09Maintains_A_Type.c09maintains_a_type) (b : bool) =
    [#"../09_maintains.rs" 25 4 25 8] true
end
module C09Maintains_Test1_Interface
  use mach.int.Int
  use mach.int.UInt64
  use C09Maintains_A_Type
  clone C09Maintains_Impl0_Invariant_Interface as Invariant0
  val test_1 [@cfg:stackify] (a : C09Maintains_A_Type.c09maintains_a_type) (b : bool) (c : uint64) : ()
    requires {[#"../09_maintains.rs" 28 0 28 31] Invariant0.invariant' a b c}
    ensures { [#"../09_maintains.rs" 28 0 28 31] Invariant0.invariant' a b c }
    
end
module C09Maintains_Test1
  use mach.int.Int
  use mach.int.UInt64
  use C09Maintains_A_Type
  clone C09Maintains_Impl0_Invariant as Invariant0
  let rec cfg test_1 [@cfg:stackify] [#"../09_maintains.rs" 29 0 29 36] (a : C09Maintains_A_Type.c09maintains_a_type) (b : bool) (c : uint64) : ()
    requires {[#"../09_maintains.rs" 28 0 28 31] Invariant0.invariant' a b c}
    ensures { [#"../09_maintains.rs" 28 0 28 31] Invariant0.invariant' a b c }
    
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : ();
  var a_1 : C09Maintains_A_Type.c09maintains_a_type;
  var b_2 : bool;
  var c_3 : uint64;
  {
    a_1 <- a;
    b_2 <- b;
    c_3 <- c;
    goto BB0
  }
  BB0 {
    _0 <- ();
    return _0
  }
  
end
module CreusotContracts_Logic_Resolve_Impl1_Resolve_Interface
  type t
  use prelude.Prelude
  predicate resolve (self : borrowed t)
end
module CreusotContracts_Logic_Resolve_Impl1_Resolve
  type t
  use prelude.Prelude
  predicate resolve (self : borrowed t) =
     ^ self =  * self
end
module CreusotContracts_Logic_Resolve_Impl1
  type t
end
module C09Maintains_Test2_Interface
  use prelude.Prelude
  use mach.int.Int
  use mach.int.UInt64
  use C09Maintains_A_Type
  clone C09Maintains_Impl0_Invariant_Interface as Invariant0
  val test_2 [@cfg:stackify] (a : borrowed (C09Maintains_A_Type.c09maintains_a_type)) (b : bool) (c : uint64) : ()
    requires {[#"../09_maintains.rs" 31 0 31 37] Invariant0.invariant' ( * a) b c}
    ensures { [#"../09_maintains.rs" 31 0 31 37] Invariant0.invariant' ( ^ a) b c }
    
end
module C09Maintains_Test2
  use prelude.Prelude
  use mach.int.Int
  use mach.int.UInt64
  use C09Maintains_A_Type
  clone C09Maintains_Impl0_Invariant as Invariant0
  clone CreusotContracts_Logic_Resolve_Impl1_Resolve as Resolve0 with type t = C09Maintains_A_Type.c09maintains_a_type
  let rec cfg test_2 [@cfg:stackify] [#"../09_maintains.rs" 32 0 32 41] (a : borrowed (C09Maintains_A_Type.c09maintains_a_type)) (b : bool) (c : uint64) : ()
    requires {[#"../09_maintains.rs" 31 0 31 37] Invariant0.invariant' ( * a) b c}
    ensures { [#"../09_maintains.rs" 31 0 31 37] Invariant0.invariant' ( ^ a) b c }
    
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : ();
  var a_1 : borrowed (C09Maintains_A_Type.c09maintains_a_type);
  var b_2 : bool;
  var c_3 : uint64;
  {
    a_1 <- a;
    b_2 <- b;
    c_3 <- c;
    goto BB0
  }
  BB0 {
    assume { Resolve0.resolve a_1 };
    _0 <- ();
    return _0
  }
  
end
module C09Maintains_Test3_Interface
  use prelude.Prelude
  use mach.int.Int
  use mach.int.UInt64
  use C09Maintains_A_Type
  clone C09Maintains_Impl0_Invariant_Interface as Invariant0
  val test_3 [@cfg:stackify] (a : borrowed (C09Maintains_A_Type.c09maintains_a_type)) (b : borrowed bool) (c : uint64) : ()
    requires {[#"../09_maintains.rs" 34 0 34 41] Invariant0.invariant' ( * a) ( * b) c}
    ensures { [#"../09_maintains.rs" 34 0 34 41] Invariant0.invariant' ( ^ a) ( ^ b) c }
    
end
module C09Maintains_Test3
  use prelude.Prelude
  use mach.int.Int
  use mach.int.UInt64
  use C09Maintains_A_Type
  clone C09Maintains_Impl0_Invariant as Invariant0
  clone CreusotContracts_Logic_Resolve_Impl1_Resolve as Resolve1 with type t = bool
  clone CreusotContracts_Logic_Resolve_Impl1_Resolve as Resolve0 with type t = C09Maintains_A_Type.c09maintains_a_type
  let rec cfg test_3 [@cfg:stackify] [#"../09_maintains.rs" 35 0 35 46] (a : borrowed (C09Maintains_A_Type.c09maintains_a_type)) (b : borrowed bool) (c : uint64) : ()
    requires {[#"../09_maintains.rs" 34 0 34 41] Invariant0.invariant' ( * a) ( * b) c}
    ensures { [#"../09_maintains.rs" 34 0 34 41] Invariant0.invariant' ( ^ a) ( ^ b) c }
    
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : ();
  var a_1 : borrowed (C09Maintains_A_Type.c09maintains_a_type);
  var b_2 : borrowed bool;
  var c_3 : uint64;
  {
    a_1 <- a;
    b_2 <- b;
    c_3 <- c;
    goto BB0
  }
  BB0 {
    assume { Resolve0.resolve a_1 };
    assume { Resolve1.resolve b_2 };
    _0 <- ();
    return _0
  }
  
end
module C09Maintains_Test5_Interface
  use mach.int.UInt64
  use mach.int.Int
  use mach.int.Int32
  use prelude.Prelude
  use C09Maintains_A_Type
  clone C09Maintains_Impl0_Inv2_Interface as Inv20
  val test_5 [@cfg:stackify] (a : C09Maintains_A_Type.c09maintains_a_type) (b : usize) : ()
    requires {[#"../09_maintains.rs" 37 0 37 28] Inv20.inv2 a (UInt64.to_int b + 0)}
    ensures { [#"../09_maintains.rs" 37 0 37 28] Inv20.inv2 a (UInt64.to_int b + 0) }
    
end
module C09Maintains_Test5
  use mach.int.UInt64
  use mach.int.Int
  use mach.int.Int32
  use prelude.Prelude
  use C09Maintains_A_Type
  clone C09Maintains_Impl0_Inv2 as Inv20
  let rec cfg test_5 [@cfg:stackify] [#"../09_maintains.rs" 38 0 38 29] (a : C09Maintains_A_Type.c09maintains_a_type) (b : usize) : ()
    requires {[#"../09_maintains.rs" 37 0 37 28] Inv20.inv2 a (UInt64.to_int b + 0)}
    ensures { [#"../09_maintains.rs" 37 0 37 28] Inv20.inv2 a (UInt64.to_int b + 0) }
    
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : ();
  var a_1 : C09Maintains_A_Type.c09maintains_a_type;
  var b_2 : usize;
  {
    a_1 <- a;
    b_2 <- b;
    goto BB0
  }
  BB0 {
    _0 <- ();
    return _0
  }
  
end
module C09Maintains_Test6_Interface
  use C09Maintains_A_Type
  clone C09Maintains_OtherInv_Interface as OtherInv0
  val test_6 [@cfg:stackify] (a : C09Maintains_A_Type.c09maintains_a_type) (b : bool) : ()
    requires {[#"../09_maintains.rs" 40 0 40 29] OtherInv0.other_inv a b}
    ensures { [#"../09_maintains.rs" 40 0 40 29] OtherInv0.other_inv a b }
    
end
module C09Maintains_Test6
  use C09Maintains_A_Type
  clone C09Maintains_OtherInv as OtherInv0
  let rec cfg test_6 [@cfg:stackify] [#"../09_maintains.rs" 41 0 41 28] (a : C09Maintains_A_Type.c09maintains_a_type) (b : bool) : ()
    requires {[#"../09_maintains.rs" 40 0 40 29] OtherInv0.other_inv a b}
    ensures { [#"../09_maintains.rs" 40 0 40 29] OtherInv0.other_inv a b }
    
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : ();
  var a_1 : C09Maintains_A_Type.c09maintains_a_type;
  var b_2 : bool;
  {
    a_1 <- a;
    b_2 <- b;
    goto BB0
  }
  BB0 {
    _0 <- ();
    return _0
  }
  
end
