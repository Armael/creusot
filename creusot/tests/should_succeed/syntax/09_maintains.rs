extern crate creusot_contracts;

use creusot_contracts::*;

// Tests that we can use field access syntax in pearlite.

struct A {}

impl A {
    #[predicate]
    fn invariant(self, b: bool, c: u64) -> bool {
        true
    }

    #[predicate]
    fn inv2(self, b: Int) -> bool {
        true
    }
}

#[predicate]
fn other_inv(a: A, b: bool) -> bool {
    true
}

#[maintains(a.invariant(b, c))]
fn test_1(a: A, b: bool, c: u64) {}

#[maintains((mut a).invariant(b, c))]
fn test_2(a: &mut A, b: bool, c: u64) {}

#[maintains((mut a).invariant(mut b, c))]
fn test_3(a: &mut A, b: &mut bool, c: u64) {}

#[maintains(a.inv2(@b + 0))]
fn test_5(a: A, b: usize) {}

#[maintains(other_inv(a, b))]
fn test_6(a: A, b: bool) {}
