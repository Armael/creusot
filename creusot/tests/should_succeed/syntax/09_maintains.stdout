module Type
  use Ref
  use mach.int.Int
  use prelude.Int8
  use prelude.Int16
  use mach.int.Int32
  use mach.int.Int64
  use prelude.UInt8
  use prelude.UInt16
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
  type c09maintains_a  = 
    | C09Maintains_A
    
end
module C09Maintains_Impl0_Invariant_Interface
  use Type
  use mach.int.Int
  use mach.int.UInt64
  predicate invariant' (self : Type.c09maintains_a) (b : bool) (c : uint64)
end
module C09Maintains_Impl0_Invariant
  use Type
  use mach.int.Int
  use mach.int.UInt64
  predicate invariant' (self : Type.c09maintains_a) (b : bool) (c : uint64) = 
    true
end
module C09Maintains_Impl0_Inv2_Interface
  use Type
  use mach.int.Int
  predicate inv2 (self : Type.c09maintains_a) (b : int)
end
module C09Maintains_Impl0_Inv2
  use Type
  use mach.int.Int
  predicate inv2 (self : Type.c09maintains_a) (b : int) = 
    true
end
module C09Maintains_OtherInv_Interface
  use Type
  predicate other_inv (a : Type.c09maintains_a) (b : bool)
end
module C09Maintains_OtherInv
  use Type
  predicate other_inv (a : Type.c09maintains_a) (b : bool) = 
    true
end
module CreusotContracts_Logic_Resolve_Impl2_Resolve_Interface
  type t
  predicate resolve (self : t)
end
module CreusotContracts_Logic_Resolve_Impl2_Resolve
  type t
  predicate resolve (self : t) = 
    true
end
module CreusotContracts_Logic_Resolve_Resolve_Resolve_Interface
  type self
  predicate resolve (self : self)
end
module CreusotContracts_Logic_Resolve_Resolve_Resolve
  type self
  predicate resolve (self : self)
end
module CreusotContracts_Logic_Resolve_Impl2
  type t
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve0 with type t = t
  clone CreusotContracts_Logic_Resolve_Resolve_Resolve as Resolve1 with type self = t,
  predicate resolve = Resolve0.resolve
end
module C09Maintains_Test1_Interface
  use Type
  use mach.int.Int
  use mach.int.UInt64
  clone C09Maintains_Impl0_Invariant_Interface as Invariant0
  val test_1 [@cfg:stackify] (a : Type.c09maintains_a) (b : bool) (c : uint64) : ()
    requires {Invariant0.invariant' a b c}
    ensures { Invariant0.invariant' a b c }
    
end
module C09Maintains_Test1
  use Type
  use mach.int.Int
  use mach.int.UInt64
  clone C09Maintains_Impl0_Invariant as Invariant0
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve2 with type t = uint64
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve1 with type t = bool
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve0 with type t = Type.c09maintains_a
  let rec cfg test_1 [@cfg:stackify] (a : Type.c09maintains_a) (b : bool) (c : uint64) : ()
    requires {Invariant0.invariant' a b c}
    ensures { Invariant0.invariant' a b c }
    
   = 
  var _0 : ();
  var a_1 : Type.c09maintains_a;
  var b_2 : bool;
  var c_3 : uint64;
  {
    a_1 <- a;
    b_2 <- b;
    c_3 <- c;
    goto BB0
  }
  BB0 {
    assume { Resolve0.resolve a_1 };
    assume { Resolve1.resolve b_2 };
    assume { Resolve2.resolve c_3 };
    _0 <- ();
    return _0
  }
  
end
module CreusotContracts_Logic_Resolve_Impl1_Resolve_Interface
  type t
  use prelude.Prelude
  predicate resolve (self : borrowed t)
end
module CreusotContracts_Logic_Resolve_Impl1_Resolve
  type t
  use prelude.Prelude
  predicate resolve (self : borrowed t) = 
     ^ self =  * self
end
module CreusotContracts_Logic_Resolve_Impl1
  type t
  use prelude.Prelude
  clone CreusotContracts_Logic_Resolve_Impl1_Resolve as Resolve0 with type t = t
  clone CreusotContracts_Logic_Resolve_Resolve_Resolve as Resolve1 with type self = borrowed t,
  predicate resolve = Resolve0.resolve
end
module C09Maintains_Test2_Interface
  use prelude.Prelude
  use Type
  use mach.int.Int
  use mach.int.UInt64
  clone C09Maintains_Impl0_Invariant_Interface as Invariant0
  val test_2 [@cfg:stackify] (a : borrowed (Type.c09maintains_a)) (b : bool) (c : uint64) : ()
    requires {Invariant0.invariant' ( * a) b c}
    ensures { Invariant0.invariant' ( ^ a) b c }
    
end
module C09Maintains_Test2
  use prelude.Prelude
  use Type
  use mach.int.Int
  use mach.int.UInt64
  clone C09Maintains_Impl0_Invariant as Invariant0
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve2 with type t = uint64
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve1 with type t = bool
  clone CreusotContracts_Logic_Resolve_Impl1_Resolve as Resolve0 with type t = Type.c09maintains_a
  let rec cfg test_2 [@cfg:stackify] (a : borrowed (Type.c09maintains_a)) (b : bool) (c : uint64) : ()
    requires {Invariant0.invariant' ( * a) b c}
    ensures { Invariant0.invariant' ( ^ a) b c }
    
   = 
  var _0 : ();
  var a_1 : borrowed (Type.c09maintains_a);
  var b_2 : bool;
  var c_3 : uint64;
  {
    a_1 <- a;
    b_2 <- b;
    c_3 <- c;
    goto BB0
  }
  BB0 {
    assume { Resolve0.resolve a_1 };
    assume { Resolve1.resolve b_2 };
    assume { Resolve2.resolve c_3 };
    _0 <- ();
    return _0
  }
  
end
module C09Maintains_Test3_Interface
  use prelude.Prelude
  use Type
  use mach.int.Int
  use mach.int.UInt64
  clone C09Maintains_Impl0_Invariant_Interface as Invariant0
  val test_3 [@cfg:stackify] (a : borrowed (Type.c09maintains_a)) (b : borrowed bool) (c : uint64) : ()
    requires {Invariant0.invariant' ( * a) ( * b) c}
    ensures { Invariant0.invariant' ( ^ a) ( ^ b) c }
    
end
module C09Maintains_Test3
  use prelude.Prelude
  use Type
  use mach.int.Int
  use mach.int.UInt64
  clone C09Maintains_Impl0_Invariant as Invariant0
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve2 with type t = uint64
  clone CreusotContracts_Logic_Resolve_Impl1_Resolve as Resolve1 with type t = bool
  clone CreusotContracts_Logic_Resolve_Impl1_Resolve as Resolve0 with type t = Type.c09maintains_a
  let rec cfg test_3 [@cfg:stackify] (a : borrowed (Type.c09maintains_a)) (b : borrowed bool) (c : uint64) : ()
    requires {Invariant0.invariant' ( * a) ( * b) c}
    ensures { Invariant0.invariant' ( ^ a) ( ^ b) c }
    
   = 
  var _0 : ();
  var a_1 : borrowed (Type.c09maintains_a);
  var b_2 : borrowed bool;
  var c_3 : uint64;
  {
    a_1 <- a;
    b_2 <- b;
    c_3 <- c;
    goto BB0
  }
  BB0 {
    assume { Resolve0.resolve a_1 };
    assume { Resolve1.resolve b_2 };
    assume { Resolve2.resolve c_3 };
    _0 <- ();
    return _0
  }
  
end
module C09Maintains_Test5_Interface
  use mach.int.UInt64
  use mach.int.Int
  use mach.int.Int32
  use Type
  use prelude.Prelude
  clone C09Maintains_Impl0_Inv2_Interface as Inv20
  val test_5 [@cfg:stackify] (a : Type.c09maintains_a) (b : usize) : ()
    requires {Inv20.inv2 a (UInt64.to_int b + 0)}
    ensures { Inv20.inv2 a (UInt64.to_int b + 0) }
    
end
module C09Maintains_Test5
  use mach.int.UInt64
  use mach.int.Int
  use mach.int.Int32
  use Type
  use prelude.Prelude
  clone C09Maintains_Impl0_Inv2 as Inv20
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve1 with type t = usize
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve0 with type t = Type.c09maintains_a
  let rec cfg test_5 [@cfg:stackify] (a : Type.c09maintains_a) (b : usize) : ()
    requires {Inv20.inv2 a (UInt64.to_int b + 0)}
    ensures { Inv20.inv2 a (UInt64.to_int b + 0) }
    
   = 
  var _0 : ();
  var a_1 : Type.c09maintains_a;
  var b_2 : usize;
  {
    a_1 <- a;
    b_2 <- b;
    goto BB0
  }
  BB0 {
    assume { Resolve0.resolve a_1 };
    assume { Resolve1.resolve b_2 };
    _0 <- ();
    return _0
  }
  
end
module C09Maintains_Test6_Interface
  use Type
  clone C09Maintains_OtherInv_Interface as OtherInv0
  val test_6 [@cfg:stackify] (a : Type.c09maintains_a) (b : bool) : ()
    requires {OtherInv0.other_inv a b}
    ensures { OtherInv0.other_inv a b }
    
end
module C09Maintains_Test6
  use Type
  clone C09Maintains_OtherInv as OtherInv0
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve1 with type t = bool
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve0 with type t = Type.c09maintains_a
  let rec cfg test_6 [@cfg:stackify] (a : Type.c09maintains_a) (b : bool) : ()
    requires {OtherInv0.other_inv a b}
    ensures { OtherInv0.other_inv a b }
    
   = 
  var _0 : ();
  var a_1 : Type.c09maintains_a;
  var b_2 : bool;
  {
    a_1 <- a;
    b_2 <- b;
    goto BB0
  }
  BB0 {
    assume { Resolve0.resolve a_1 };
    assume { Resolve1.resolve b_2 };
    _0 <- ();
    return _0
  }
  
end
